File: typescript/src/deliveryController.ts
import {EmailGateway} from './emailGateway'
import {Location, MapService} from './mapService';

const TEN_MINUTES = 1000 * 60 * 10;

export interface Delivery {
   id: string
    contactEmail: string
    location: Location
    timeOfDelivery: Date
    arrived: boolean
    onTime: boolean
}

export interface DeliveryEvent {
    id: string
    timeOfDelivery: Date
    location: Location
}

export class DeliveryController {
    #emailGateway: EmailGateway;
    #mapService: MapService;
    #deliveries: Array<Delivery>;

    constructor(deliveries: Array<Delivery>) {
        this.#deliveries = deliveries;
        this.#mapService = new MapService();
        this.#emailGateway = new EmailGateway();
    }

    public async updateDelivery(event: DeliveryEvent) {
        let nextDelivery: Delivery;

        for(let i = 0; i < this.#deliveries.length; i++){
            const delivery = this.#deliveries[i];
            if (delivery.id === event.id) {
                delivery.arrived = true;
                var timeDifference = delivery.timeOfDelivery.getTime() - event.timeOfDelivery.getTime();
                if (timeDifference < TEN_MINUTES) {
                    delivery.onTime = true;
                }
                delivery.timeOfDelivery = event.timeOfDelivery;
                let message = `Regarding your delivery today at ${delivery.timeOfDelivery}. How likely would you be to recommend this delivery service to a friend? Click <a href='url'>here</a>`
                this.#emailGateway.send(delivery.contactEmail, "Your feedback is important to us", message)
                if(this.#deliveries.length > i + 1) {
                    nextDelivery = this.#deliveries[i + 1];
                }

                if(!delivery.onTime && this.#deliveries.length > 1 && i > 0) {
                    var previousDelivery = this.#deliveries[i - 1];
                    var elapsedTime = delivery.timeOfDelivery.getTime() - previousDelivery.timeOfDelivery.getTime();
                    this.#mapService.updateAverageSpeed(previousDelivery.location, delivery.location, elapsedTime);
                }
            }
        }

        if (nextDelivery !== undefined) {
            var nextEta = this.#mapService.calculateETA(event.location, nextDelivery.location);
            const message = `Your delivery to ${nextDelivery.location} is next, estimated time of arrival is in ${nextEta} minutes. Be ready!`
            await this.#emailGateway.send(nextDelivery.contactEmail, "Your delivery will arrive soon.", message);
        }
    }
}

File: python/delivery_controller.py
import datetime
from dataclasses import dataclass

from email_gateway import EmailGateway
from map_service import MapService, Location


@dataclass(frozen=True)
class DeliveryEvent:
    id: str
    time_of_delivery: datetime.datetime
    location: Location

@dataclass
class Delivery:
    id: str
    contact_email: str
    location: Location
    time_of_delivery: datetime.datetime
    arrived: bool
    on_time: bool


class DeliveryController:

    def __init__(self, delivery_schedule : list):
        self.delivery_schedule = delivery_schedule
        self.email_gateway = EmailGateway()
        self.map_service = MapService()

    def update_delivery(self, delivery_event : DeliveryEvent):
        next_delivery = None
        for i, delivery in enumerate(self.delivery_schedule):
            if delivery_event.id == delivery.id:
                delivery.arrived = True
                time_difference = delivery_event.time_of_delivery - delivery.time_of_delivery
                if time_difference < datetime.timedelta(minutes=10):
                    delivery.on_time = True
                delivery.time_of_delivery = delivery_event.time_of_delivery
                message = f"""Regarding your delivery today at {delivery.time_of_delivery}. How likely would you be to recommend this delivery service to a friend? Click <a href="url">here</a>"""
                self.email_gateway.send(delivery.contact_email, "Your feedback is important to us", message)
                if len(self.delivery_schedule) > i+1:
                    next_delivery = self.delivery_schedule[i+1]
                if not delivery.on_time and len(self.delivery_schedule) > 1 and i > 0:
                    previous_delivery = self.delivery_schedule[i-1]
                    elapsed_time = delivery.time_of_delivery - previous_delivery.time_of_delivery
                    self.map_service.update_average_speed(previous_delivery.location, delivery.location, elapsed_time)
        if next_delivery:
            next_eta = self.map_service.calculate_eta(delivery_event.location, next_delivery.location)
            message = f"Your delivery to {next_delivery.location} is next, estimated time of arrival is in {next_eta:.0f} minutes. Be ready!"
            self.email_gateway.send(next_delivery.contact_email, "Your delivery will arrive soon", message)



